### PR TITLE
[FLINK-26726][connector][hive]Hive enumerators do not assign splits to unregistered (failed) readers.

### PR DESCRIPTION
**Pull request by zoucao**
## What is the purpose of the change

  Let hive-enumerators do not assign splits to unregistered (failed) readers.

## Brief change log

  - add filter logic in ContinuousHiveSplitEnumerator#assignSplits
  - add test for ContinuousHiveSplitEnumerator

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluster with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zoucao** [comment]
@flinkbot run azure

**Comment by zoucao** [comment]
@flinkbot run azure

**Comment by zoucao** [comment]
Gentle ping @JingsongLi, could you help me to review it?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+/** Unit tests for the {@link ContinuousHiveSplitEnumerator}. */
+public class ContinuousHiveSplitEnumeratorTest {
+
+    @Test
+    public void testDiscoverSplitWhenNoReaderRegistered() throws Exception {
+        final TestingSplitEnumeratorContext<HiveSourceSplit> context =
+                new TestingSplitEnumeratorContext<>(4);
+
+        final ContinuousHiveSplitEnumerator enumerator =
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:55]: nit:
```suggestion
        final ContinuousHiveSplitEnumerator<Long> enumerator =
```

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                new ContinuousHiveSplitEnumerator(
+                        context,
+                        0L,
+                        Collections.emptySet(),
+                        new SimpleSplitAssigner(Collections.emptyList()),
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:60]: How about using 'new SimpleSplitAssigner(Collections.singletonList(split)),'?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        assertEquals(1, enumerator.snapshotState(1L).getSplits().size());
+        assertTrue(enumerator.snapshotState(1L).getSplits().contains(split));
+    }
+
+    @Test
+    public void testDiscoverWhenReaderRegistered() throws Exception {
+        final TestingSplitEnumeratorContext<HiveSourceSplit> context =
+                new TestingSplitEnumeratorContext<>(4);
+
+        final ContinuousHiveSplitEnumerator enumerator =
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:82]: dito

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        assertEquals(0, enumerator.snapshotState(1L).getSplits().size());
+        assertTrue(context.getSplitAssignments().get(2).getAssignedSplits().contains(split));
+    }
+
+    @Test
+    public void testRequestingReaderUnavailableWhenSplitDiscovered() throws Exception {
+        final TestingSplitEnumeratorContext<HiveSourceSplit> context =
+                new TestingSplitEnumeratorContext<>(4);
+
+        final ContinuousHiveSplitEnumerator enumerator =
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:112]: dito

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                0,
+                0,
+                0,
+                0,
+                new String[] {"host1"},
+                null,
+                new HiveTablePartition(sd, new HashMap<>(), new Properties()));
+    }
+
+    private ContinuousPartitionFetcher mockPartitionFetcher() {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:155]: nit
```suggestion
    private ContinuousPartitionFetcher<Partition, Long> mockPartitionFetcher() {
```

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+            @Override
+            public List<Partition> fetch(PartitionFetcher.Context<Partition> context)
+                    throws Exception {
+                return Collections.emptyList();
+            }
+        };
+    }
+
+    private class MockHiveContinuousPartitionFetcherContext
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:174]:
```suggestion
    private static class MockHiveContinuousPartitionFetcherContext
```

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+            extends HiveTableSource.HiveContinuousPartitionFetcherContext {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:175]:
```suggestion
            extends HiveTableSource.HiveContinuousPartitionFetcherContext<Long> {
```

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        }
+
+        @Override
+        public void close() throws Exception {}
+
+        @Override
+        public void open() throws Exception {}
+
+        @Override
+        public Optional<Partition> getPartition(List partValues) throws TException {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:188]:
```suggestion
        public Optional<Partition> getPartition(List<String> partValues) throws TException {
```

**Review by luoyuxia — COMMENTED** [review]
@zoucao Thanks for contribution. I left some comments.
And could you please rebase master? and then I'll find someone to merge your pr.

**Comment by zoucao** [comment]
> 

Sure, I will rebase the master and fix it as soon as possible.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+                        new ObjectPath("testDb", "testTable"),
+                        mockPartitionFetcher(),
+                        new MockHiveContinuousPartitionFetcherContext(
+                                new ObjectPath("testDb", "testTable")));
+        enumerator.start();
+
+        HiveSourceSplit split = createSplit();
+        enumerator.addSplitsBack(Collections.singletonList(split), 0);
+        context.triggerAllActions();
+        assertEquals(1, enumerator.snapshotState(1L).getSplits().size());
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumeratorTest.java:73]: Sorry for missing this line. Please use `assertThat` just like what `ContinuousFileSplitEnumeratorTest` does.

**Comment by zoucao** [comment]
@flinkbot run azure

**Comment by zoucao** [comment]
> @zoucao Thanks for updating. Could you please open the same pr to release-1.16/1.15/1.14? @wuchong Could you please help merge?
Yes, I will do the bp soon.

**Comment by zoucao** [comment]
@luoyuxia plz help to confirm it.
to 1.16 https://github.com/apache/flink/pull/21036
to 1.15 https://github.com/apache/flink/pull/21038
to 1.14 https://github.com/apache/flink/pull/21044

**Comment by luoyuxia** [comment]
@zoucao Thanks for efforts.

**Pull request merged**